[PyTorch] Make TensorImpl::sizes() customizable and disable it for NestedTensorImpl

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73817

NestedTensorImpl doesn't have sizes(). Silently getting wrong results back from it is not conducive to efficient software development. Make it throw while allowing sizes() to be inlined in the common case anyway, just like is_contiguous(). Thanks @ezyang for the reminder that we could do this.

Differential Revision: [D34660829](https://our.internmc.facebook.com/intern/diff/D34660829/)

swolchok committed:
[PyTorch] Make TensorImpl::sizes() customizable and disable it for NestedTensorImpl

NestedTensorImpl doesn't have sizes(). Silently getting wrong results back from it is not conducive to efficient software development. Make it throw while allowing sizes() to be inlined in the common case anyway, just like is_contiguous(). Thanks @ezyang for the reminder that we could do this.

Differential Revision: [D34660829](https://our.internmc.facebook.com/intern/diff/D34660829/)

[ghstack-poisoned]

ezyang reviewed (approved):
Approving for speed. I'm wondering how much we actually want to use two bits on the policy; maybe we can do the error case by just going to the dynamic dispatch (and then having that raise the error).

swolchok commented:
> Approving for speed. I'm wondering how much we actually want to use two bits on the policy; maybe we can do the error case by just going to the dynamic dispatch (and then having that raise the error).

I don't think we have a use case for the dynamic dispatch yet, which is why it's not even implemented. I could make it 1 bit but I think that would raise bc issues if we add the dynamic dispatch? We can always make the storage 1 bit despite having the 3 cases in the policy if we need the bit back.

swolchok committed:
Update on "[PyTorch] Make TensorImpl::sizes() customizable and disable it for NestedTensorImpl"


NestedTensorImpl doesn't have sizes(). Silently getting wrong results back from it is not conducive to efficient software development. Make it throw while allowing sizes() to be inlined in the common case anyway, just like is_contiguous(). Thanks ezyang for the reminder that we could do this.

Differential Revision: [D34660829](https://our.internmc.facebook.com/intern/diff/D34660829/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make TensorImpl::sizes() customizable and disable it for NestedTensorImpl"


NestedTensorImpl doesn't have sizes(). Silently getting wrong results back from it is not conducive to efficient software development. Make it throw while allowing sizes() to be inlined in the common case anyway, just like is_contiguous(). Thanks ezyang for the reminder that we could do this.

Differential Revision: [D34660829](https://our.internmc.facebook.com/intern/diff/D34660829/)

[ghstack-poisoned]
